File: HackerRank/Dynamic_Programming/The_Coin_Change_Problem.go
package main

import (
	"fmt"
	"os"
	"bufio"
)

func main() {
	reader := bufio.NewReader(os.Stdin)
	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()

	var n, m int
	fmt.Fscanln(reader, &n, &m)
	var coins = make([]int, m)
	for i := range(coins) {
		fmt.Fscanf(reader, "%d ", &coins[i])
	}

	fmt.Fprintln(writer, getWays(n, coins))
}

func getWays(n int, coins []int) int {
	var ways = make([]int, n+1)
	ways[0] = 1
	for _, coin := range(coins) {
		for i := 1; i<n+1; i++ {
			if coin <= i {
				ways[i] += ways[i-coin]
			}
		}
	}
	return ways[n]
}

File: HackerRank/Dynamic_Programming/The_Coin_Change_Problem.py
import sys

def get_ways(n, coins):
    ways = [0]*(n+1)
    ways[0] = 1
    for _, coin in enumerate(coins):
        for i in range(1, n+1):
            if coin <= i:
                ways[i] += ways[i-coin]
    return ways[n]

if __name__ == "__main__":
    n, m = map(int, sys.stdin.readline().split())
    coins = list(map(int, sys.stdin.readline().split()))
    print(get_ways(n, coins))